nmigen/vendor/altera.py

```diff
@@ -121,45 +121,140 @@ def _invert_if(invert, value):
 
     def get_input(self, pin, port, attrs, invert):
         self._check_feature("single-ended input", pin, attrs,
-                            valid_xdrs=(0,), valid_attrs=True)
+                            valid_xdrs=(0,1), valid_attrs=True)
 
         m = Module()
-        m.d.comb += pin.i.eq(self._invert_if(invert, port))
+
+        ff_i = Signal(pin.width)
+
+        for bit in range(pin.width):
+            m.submodules["{}_buf_{}".format(pin.name, bit)] = Instance("alt_inbuf",
+                i_i=port[bit],
+                o_o=ff_i[bit]
+            )
+            
+            if pin.xdr == 0:
+                m.d.comb += pin.i[bit].eq(self._invert_if(invert, ff_i[bit]))
+            elif pin.xdr == 1:
+                m.submodules["{}_dff_i_{}".format(pin.name, bit)] += Instance("dff",
+                    i_d=self._invert_if(invert, ff_i[bit]),
+                    i_clk=pin.i_clk,
+                    i_clrn=1,
+                    i_prn=1,
+                    o_q=pin.i[bit]
+                )
+
         return m
```

Comment:
According [AN307](https://www.intel.com/content/dam/www/programmable/us/en/pdfs/literature/an/an307.pdf), table 56 you should probably set the attribute `FAST_INPUT_REGISTER` to use the reg within the IOB.